### DISCUSSION TITLE
How to unlock an iModel Briefcase

### DISCUSSION THREAD

**Discussion by PeterJBassett** [iTwin/itwinjs-core#3938] [Q&A] [answered]
Hello

I believe it is the case that internal Bentley developers have access to an easy to use UI to unlock an iModel briefcase that has been left in a locked state by an error.

However, I have been informed that this unlock UI is not available to external developers :( Instead I have been told to use an API call to perform unlocking ( https://imodelhubapi.bentley.com/swagger/index.html ).

Looking at the API calls available, it is not clear which API call to use and how to unlock a locked briefcase. Which Lock API to use and which API request to use (POST, DELETE ?)

Please could you provide some guidance?

Finally, why cant external developers be given the same easy to use UI for unlocking iModels that internal Bentley developers have access to? Personnaly, developing the FutureOn to iTwin connector, I have had to spend a lot of unecessary time deleting and re-creating iModels that have become locked.

**Comment by pmconne** [comment, marked as answer]
@jffmarker @abeesh @timlawrence-bentley

**Reply by jffmarker** [reply]
If I had to guess from the swagger, I'd look at DELETE ​/sv1.1​/Repositories​/iModel--{iModelId}​/iModelScope​/Lock​/{instanceId}. It has docs on how to format the instanceId based on what you want to unlock. I'll also forward this discussion to someone on our Hub team to try to help.

**Reply by aerimagne** [reply]
If you are having an issue with briefcase owning locks and not releasing them you can either release locks manually or release the briefcase. If you're releasing locks, you would have to query and update locks across multiple requests. If you release briefcase, it can be done in a single request, it would also release codes (if you're using them), but it can potentially timeout (if you are releasing a large amount of locks) and you would have to redownload the briefcase.

I would suggest to use either iModels APIs ([release briefcase](https://developer.bentley.com/apis/imodels/operations/release-imodel-briefcase/), [get](https://developer.bentley.com/apis/imodels/operations/get-imodel-locks/) / [update locks](https://developer.bentley.com/apis/imodels/operations/update-imodel-locks/)) or iModels clients [npm package](https://www.npmjs.com/package/@itwin/imodels-client-authoring) ([release briefcase](https://github.com/iTwin/imodels-clients/blob/main/clients/imodels-client-authoring/src/operations/briefcase/BriefcaseOperations.ts#L40) / [update locks](https://github.com/iTwin/imodels-clients/blob/main/clients/imodels-client-authoring/src/operations/lock/LockOperations.ts#L39)). If you're on 2.19.x, there is [deleteAll](https://github.com/iTwin/itwinjs-core/blob/9ba30e31a810594776731ebe06b6af88f35a51bc/clients/imodelhub/src/imodelhub/Locks.ts#L549) method for locks, but that client has been replaced by iModels clients.

**Reply by PeterJBassett** [reply]
Hello @Animagne 
I have tried to use the API GET / PATCH identified in your reply, but this does not seem to work.
If I run the GET request with my iModel Id the response is:
```
{
    "locks": [],
    "_links": {
        "self": {
            "href": "https://api.bentley.com/imodels/c883d2e5-f473-469b-8d53-851635983cb1/locks?$skip=0&$top=100"
        },
        "prev": null,
        "next": null
    }
}
```

This implies that the iModel is not locked. However, when I try to run a synchronisation, I get the error message:
```
  errorNumber: 102428,
  _getMetaData: undefined,
  description: 'cannot POST /imodelhub/sv1.1/Repositories/iModel--c883d2e5-f473-469b-8d53-851635983cb1/$changeset (409)',
  _data: {
    errorId: 'iModelHub.LockOwnedByAnotherBriefcase',
    errorMessage: 'Lock(s) is owned by another briefcase.',
    ConflictingLocks: [ [Object] ]
  },
  status: 409,
  data: {
    errorId: 'iModelHub.LockOwnedByAnotherBriefcase',
    errorMessage: 'Lock(s) is owned by another briefcase.',
    ConflictingLocks: [ [Object] ]
  }
}

```

**Reply by aerimagne** [reply]
The request seems to be correct and it should have returned some locks. It will need more investigation why no locks were returned to you.

You should now have access to release locks through iModel Manager website in connect portal ("Manage resources" should appear under settings icon). I don't know if this is internal GUI you were referring to, but hopefully this should help solving these lock issues.

**Reply by PeterJBassett** [reply]
Hello - Thank you for your help. There is no "Manage resources" option available?
<img width="1017" alt="Screenshot 2022-07-21 at 12 44 30" src="https://user-images.githubusercontent.com/61423714/180206204-3dabaa62-2a94-40c8-a0c5-3c066959121d.png">

**Reply by pmconne** [reply]
Locks are managed per-iModel. From the iModel Manager page shown above, select the specific iModel you want to unlock.
![image](https://user-images.githubusercontent.com/22944042/180207867-c9270fab-2630-49c0-a6e7-8f713fd80b27.png)

**Reply by PeterJBassett** [reply]
Ah ok - that is exactly the UI I was hoping for. I was told on a few occassions it wasn't available to external (non-Bentley) users.

Thanks very much.

**Reply by PeterJBassett** [reply]
Hello @pmconne - However, when I try to release locks it doesn't work. It gives the message "0 locks released successfully"

**Reply by pmconne** [reply]
You're going to need @Animagne's assistance there.

**Reply by aerimagne** [reply]
This UI allows releasing locks for individual briefcase. I saw that you have tried to release locks for couple of briefcases (based on logs, I think 88 and 150). As I understand, normally when connectors end they run, they are supposed to release any lock they acquired, so it would be normal for these briefcases to no longer own any locks.

Since this iModel is on Bentley organization, I was able to query locks using the API document in that swagger page and find that Briefcase 146 is currently holding a lock.

**Reply by PeterJBassett** [reply]
Hello @Animagne - the problem is that there can be abnormal cases. If a connector synchronisation fails due to an error, it can leave an iModel locked, so that it can no longer be used, or the resynchronisation re-tried.

This abnormal sitruation can happen occassionally. Therefore there is a need to be able to unlock the iModel.

So - is the answer, that I need to run the API query to find out which brielcase is locked, and then use the UI to unlock that briefacse?

Which API query did you use? 

Does this API work for external (non-Bentley) users?

**Reply by pmconne** [reply]
> If a connector synchronisation fails due to an error, it can leave an iModel locked

This sounds like another negative consequence of the connector acquiring a new briefcase for each synchronization (as discussed in #3951), leaving the previous briefcase hanging around with its locks intact. Ideally the connector would reuse the same briefcase for each synchronization. Less ideally, it would at least release previously-acquired briefcases (and their locks) on the next synchronization.

**Reply by aerimagne** [reply]
You can either query to see what briefcases are holding locks or look into the iModelHub.LockOwnedByAnotherBriefcase error message (ConflictingLocks object there should include a list of locks and briefcases ids that own them).

I've used GET https://imodelhubapi.bentley.com/sv1.1/Repositories/iModel--{iModelId}/iModelScope/Lock?$filter=LockLevel+ne+0 (just added the filter to the request available in swagger to only show currently held locks). This API is available for external users, but not recommended for new applications as it is being replaced by the APIs documented in https://developer.bentley.com/apis/imodels/operations. I have reproduced the issue with the latter API, where it didn't return any locks, and we will be working on a fix for it.
